core/src/storage2/collections/hashmap/mod.rs

```diff
@@ -337,4 +384,204 @@ where
+where
+    K: Ord + Clone + PackedLayout,
+    V: PackedLayout + core::fmt::Debug + core::cmp::Eq + Default,
+    H: Hasher,
+    Key: From<<H as Hasher>::Output>,
+{
+    /// Returns a reference to this entry's key.
+    pub fn key(&self) -> &K {
+        match self {
+            Entry::Occupied(entry) => &entry.key,
+            Entry::Vacant(entry) => &entry.key,
+        }
+    }
+
+    /// Ensures a value is in the entry by inserting the default value if empty, and returns
+    /// a reference to the value in the entry.
+    pub fn or_default(self) -> &'a V {
+        match self {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => entry.insert(V::default()),
+        }
+    }
+
+    /// Ensures a value is in the entry by inserting the default if empty, and returns
+    /// a mutable reference to the value in the entry.
+    pub fn or_insert(self, default: V) -> &'a mut V {
+        match self {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => entry.insert(default),
+        }
+    }
+
+    /// Ensures a value is in the entry by inserting the result of the default function if empty,
+    /// and returns mutable references to the key and value in the entry.
+    pub fn or_insert_with<F>(self, default: F) -> &'a mut V
+    where
+        F: FnOnce() -> V,
+    {
+        match self {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => Entry::insert(default(), entry),
+        }
+    }
+
+    /// Ensures a value is in the entry by inserting, if empty, the result of the default
+    /// function, which takes the key as its argument, and returns a mutable reference to
+    /// the value in the entry.
+    pub fn or_insert_with_key<F>(self, default: F) -> &'a mut V
+    where
+        F: FnOnce(&K) -> V,
+    {
+        match self {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => Entry::insert(default(&entry.key), entry),
+        }
+    }
+
+    /// Provides in-place mutable access to an occupied entry before any
+    /// potential inserts into the map.
+    pub fn and_modify<F>(self, f: F) -> Self
+    where
+        F: FnOnce(&mut V),
+    {
+        match self {
+            Entry::Occupied(mut entry) => {
+                {
+                    let v = entry.get_mut();
+                    f(v);
+                }
+                Entry::Occupied(entry)
+            }
+            Entry::Vacant(entry) => Entry::Vacant(entry),
+        }
+    }
+
+    /// Inserts `value` into `entry`.
+    fn insert(value: V, entry: VacantEntry<'a, K, V, H>) -> &'a mut V {
+        let old_value = entry.base.insert(entry.key.clone(), value);
+        debug_assert!(old_value.is_none());
+        match entry.base.entry(entry.key) {
+            Entry::Vacant(_) => unreachable!("entry was just inserted; qed"),
+            Entry::Occupied(entry) => entry.into_mut(),
+        }
```

Comment:
Couldn't this be replaced by `entry.base.get_mut(entry_key).expect("encountered invalid vacant entry")` ?